projectq/types/_qubit.py

```diff
@@ -116,12 +116,18 @@ class Qubit(BasicQubit):
     Thus the qubit is not copyable; only returns a reference to the same
     object.
     """
+    def __init__(self, engine, idx):
+        BasicQubit.__init__(self, engine, idx)
+
     def __del__(self):
         """
         Destroy the qubit and deallocate it (automatically).
         """
-        self.engine.deallocate_qubit(self)
+        if self.id == -1:
```

Comment:
There are already engines in the codebase with a `deallocate_qubit` method that don't do this check (in Simulator and in a mock test engine). This check reduces a complicated whole-code-base proof, about what's a main engine and what's a backend and how they get called, into a one-liner.

(... If it was up to me I'd make "you have to give me a valid qubit" a precondition of `engine.deallocate_qubit` and throw when it happened. Double-free's are symptoms of bugs.)